### PR TITLE
ENH: add ctree dist app

### LINKED ISSUE #52: ENH: provide mash distance as a standalone app
`dvs_mash` should take `c3_types.SeqsCollectionType` as the input type and return `c3_types.PairwiseDistanceType`

### PR DESCRIPTION
**Pull request by rmcar17**
Stacked on #56 

Resolves #52

### CODE REVIEW

File: src/diverse_seq/distance.py [hunk 1]
```diff
@@ -1,28 +1,473 @@
...
 
 
-@_intersect_union.register(set)
-def _(rec1: set, rec2: set) -> tuple[int, int]:
-    intersect = rec1 & rec2
-    union = rec1 | rec2
-    return len(intersect), len(union)
+@define_app
+class dvs_dist:
+    """return the N most divergent sequences"""
```
Comment (GavinHuttley) [inline src/diverse_seq/distance.py:24]: wrong docstring

File: src/diverse_seq/distance.py [hunk 2]
```diff
@@ -1,28 +1,473 @@
...
+        """Initialise parameters for kmer distance calculation.
 
-@_intersect_union.register(ndarray)
-def _(rec1: ndarray, rec2: ndarray) -> tuple[int, int]:
-    intersect = intersect1d(rec1, rec2, assume_unique=True)
-    union = union1d(rec1.data, rec2.data)
-    return len(intersect), len(union)
+        Parameters
+        ----------
+        distance_mode : Literal[&quot;mash&quot;, &quot;euclidean&quot;], optional
```
Comment (GavinHuttley) [inline src/diverse_seq/distance.py:40]: remove html escapes, you can use double quotes within docstrings. My pref is leave the type hints to the python code

File: src/diverse_seq/distance.py [hunk 3]
```diff
@@ -1,28 +1,473 @@
...
+                    moltype=self._moltype,
+                )
+                for seq in seq_arrays
+            ]
+            distances = euclidean_distances(
+                kmer_seqs,
+                with_progress=self._with_progress,
+            )
+        else:
+            msg = f"Unexpected distance {self._distance_mode}."
```
Comment (GavinHuttley) [inline src/diverse_seq/distance.py:121]: should be checked in constructor

**Review by GavinHuttley — COMMENTED** [review]
this looks really similar to one of your other PR's?

File: src/diverse_seq/distance.py (hunk 2 above)
Comment (rmcar17) [reply src/diverse_seq/distance.py:40]: ah, my automatic docstring thing put them in automatically (as with the type hints - can remove them in the docstring if you'd like me to).

File: src/diverse_seq/distance.py [hunk 4]
```diff
@@ -65,6 +68,10 @@ def __init__(
         if mash_canonical_kmers is None:
             mash_canonical_kmers = False
 
+        if distance_mode not in ("mash", "euclidean"):
+            msg = f"Unexpected distance {distance_mode}."
```
Comment (GavinHuttley) [inline src/diverse_seq/distance.py:72]: change `f"Unexpected distance {distance_mode}."` to `f"Unexpected distance {distance_mode!r}."` Makes the error message easier to read by using the representation.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
Tiny changes. Can you add the `# pylint: disable=not-callable` comments to address the  two codacy issues. Also, I do prefer not having type hints in the docstring since the function / method signature contains them

**Comment by rmcar17** [comment]
I'm beginning to get rebasing conflicts between the branches so I'll close this pull request and propagate the changes to the other one.

**Pull request closed without merging**